Lib/asyncio/runners.py

```diff
@@ -1,10 +1,85 @@
-__all__ = 'run',
+__all__ = ('Runner', 'run')
 
+import contextvars
 from . import coroutines
 from . import events
 from . import tasks
 
 
+class Runner:
+    """A context manager that controls event loop life cycle.
+
+    The context manager always creates a new event loop,
+    allows to run async functions inside it,
+    and properly finalizes the loop at the context manager exit.
+
+    If debug is True, the event loop will be run in debug mode.
+    If factory is passed, it is used for new event loop creation.
+
+    asyncio.run(main(), debug=True)
+
+    is a shortcut for
+
+    with asyncio.Runner(debug=True) as runner:
+        runner.run(main())
+
+    The run() method can be called multiple times within the runner's context.
+
+    This can be useful for interactive console (e.g. IPython),
+    unittest runners, console tools, -- everywhere when async code
+    is called from existing sync framework and where the preferred single
+    asyncio.run() call doesn't work.
+
+    """
+    def __init__(self, *, debug=None, factory=None):
+        if factory is None:
+            self._loop = events.new_event_loop()
+        else:
+            self._loop = factory()
+        if debug is not None:
+            self._loop.set_debug(debug)
+        self._context = contextvars.copy_context()
```

Comment:
I'm not sure that the context should be captured when you create a Runner object. Maybe it should be captured for the first time when you call `__enter__()` or `.run()`? IDK.

Reply:
Runner constructor and `__enter__` are usually placed in the same line.
But sometimes `__enter__`/`__exit__` pair is not used, `runner.close()` is called explicitly.

Moving initialization to the first `run()` call is possible but complicates the code a little.
What is the point? Do you have a use case for postponed evaluation?

I can imagine when the postponing can be useful for, say, `asyncio.Queue` or an asyncio HTTP client. But the runner is used exactly after creation, isn't it?

Reply:
> I can imagine when the postponing can be useful for, say, asyncio.Queue or an asyncio HTTP client. But the runner is used exactly after creation, isn't it?

It should be, but what if it's not? This isn't a big issue or anything, I just want to have a quick discussion here.

Why do we need to capture the context at all, and not just let it use whatever context was there when the loop is about to start running? Perhaps we don't need to add `context` to the runner at all? Or maybe add it to `run()` only and set it to `None` by default?

Reply:
1. Consider the following:
```
async with Runner() as runner:
    runner.run(coro1())
    runner.run(coro2())
```
Without the context capturing `coro2` doesn't see context modifications done by `coro1`. I've found it confusing.

2. Support `context` argument is useful for test case at least: https://github.com/python/cpython/pull/31799/files#diff-1f2ae0f6c6010caf9d5f1c80cd6033a796ffe2b60554f5df84f554f4a08e622bR80-R83
If we have no consensus here I can replace the single call with 
```
task = runner.get_loop().create_task(coro(), context=context)
runner.run(task)
```
Looks a little complicated but I can live with it, sure.

Please note, `pytest-asyncio` cannot reuse `context` from the previous `run` call but should combine the current context from all contexts from preceding fixtures.

P.S.
If you are in doubt, we can drop contexts entirely now and add them in future Python versions.  The change looks *almost* backward compatible.

Reply:
Hmm, `runner.run()` should accept `context` optional argument.
Otherwise, it should accept a task as the first argument (now the only coroutine is allowed which I like very much).

Reply:
Rewritten with capture-on-first-call.